### PR TITLE
Add record and replay helpers

### PR DESCRIPTION
**Pull request by npmitche**

### CODE REVIEW

File: chipsec/helper/record/recordhelper.py [hunk 1]
```diff
@@ -0,0 +1,335 @@
...
+
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+
+# You should have received a copy of the GNU General Public License
+# along with this program; if not, write to the Free Software
+# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+
```
Comment (frinzell) [inline chipsec/helper/record/recordhelper.py:15]: Missing copyright year

File: chipsec/helper/record/recordhelper.py [hunk 2]
```diff
@@ -0,0 +1,335 @@
...
+        ret = self._subhelper.write_phys_mem(phys_address, length, buf)
+        self._add_element("write_phys_mem", (phys_address, length, buf), ret)
+        return ret
+
+    def alloc_phys_mem(self, length: int, max_phys_address: int) -> Tuple[int, int]:
+        ret = self._subhelper.alloc_phys_mem(length, max_phys_address)
+        self._add_element("alloc_phys_mem", (length, max_phys_address), ret)
+        return ret
+
+    def free_phys_mem(self, physical_address: int):
```
Comment (frinzell) [inline chipsec/helper/record/recordhelper.py:150]: Missing return type

File: chipsec/helper/record/recordhelper.py [hunk 3]
```diff
@@ -0,0 +1,335 @@
...
+        
+        dtstr = datetime.now().strftime("%Y%m%d%H%M%S")
+        self.default_file_name = f"recording{dtstr}.json"
+        self.default_file_location = os.path.join("chipsec", "helper", "record")
+        self._subhelper = helper().get_helper(subhelper) if subhelper else helper().get_default_helper()
+        logger().log(f"Using subhelper: {self._subhelper.name}")
+        self._filename = filename if filename else os.path.join(self.default_file_location, self.default_file_name)
+        self._data = {}
+
+    def switch_subhelper(self, newhelper: Helper):
```
Comment (frinzell) [inline chipsec/helper/record/recordhelper.py:52]: Return type

File: chipsec/helper/record/recordhelper.py [hunk 4]
```diff
@@ -0,0 +1,335 @@
...
+        if self.driver_loaded:
+            self._subhelper.stop()
+        if self.driver_created:
+            self._subhelper.delete()
+        self._subhelper = newhelper
+        logger().log(f"Switching subhelper to: {self._subhelper.name}")
+        self.create()
+        self.start()
+
+    def _add_element(self, cmd, args, ret):
```
Comment (frinzell) [inline chipsec/helper/record/recordhelper.py:62]: Types

File: chipsec/helper/record/recordhelper.py [hunk 5]
```diff
@@ -0,0 +1,335 @@
...
+from chipsec.helper.basehelper import Helper
+from chipsec.helper.oshelper import helper
+from chipsec.file import read_file, write_file
+
+
+class RecordHelper(Helper):
+    """Default test helper that emulates a Broadwell architecture.
+    See datasheet for registers definition.
+    http://www.intel.com/content/www/us/en/chipsets/9-series-chipset-pch-datasheet.html
+    """
```
Comment (frinzell) [inline chipsec/helper/record/recordhelper.py:32]: Is this doc string applicable with this class?

File: chipsec/helper/replay/replayhelper.py [hunk 6]
```diff
@@ -0,0 +1,248 @@
...
+
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+
+# You should have received a copy of the GNU General Public License
+# along with this program; if not, write to the Free Software
+# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+
```
Comment (frinzell) [inline chipsec/helper/replay/replayhelper.py:15]: Copyright year

File: chipsec/helper/replay/replayhelper.py [hunk 7]
```diff
@@ -0,0 +1,248 @@
...
+from chipsec.file import read_file
+from chipsec.logger import logger
+from chipsec.helper.basehelper import Helper
+
+
+class ReplayHelper(Helper):
+    """Default test helper that emulates a Broadwell architecture.
+    See datasheet for registers definition.
+    http://www.intel.com/content/www/us/en/chipsets/9-series-chipset-pch-datasheet.html
+    """
```
Comment (frinzell) [inline chipsec/helper/replay/replayhelper.py:32]: Is this applicable?

File: chipsec/helper/replay/replayhelper.py [hunk 8]
```diff
@@ -0,0 +1,248 @@
...
+        file_data = read_file(self.config_file)
+        if file_data == 0:
+            raise OsHelperError(f"Unable to open JSON File: {self.config_file}", EACCES)
+        try:
+            self._data = loads(file_data)
+        except Exception:
+            raise OsHelperError(f'Unable to load JSON File: {self.config_file}', EFAULT)
+
+
+    # This will be used to probe the device, fake a Broadwell CPU
```
Comment (frinzell) [inline chipsec/helper/replay/replayhelper.py:96]: Needed comment?

File: chipsec/helper/replay/replayhelper.py [hunk 9]
```diff
@@ -0,0 +1,248 @@
...
+        except Exception:
+            return None
+        return evaledobject
+
+    def _get_element(self, cmd: str, args: Tuple) -> Optional[Any]:
+        try:
+            targs = '({})'.format(','.join(str(i) for i in args))
+        except Exception:
+            targs = str(args)
+        # margs = targs.encode('latin_1')
```
Comment (frinzell) [inline chipsec/helper/replay/replayhelper.py:79]: Dead code.

File: tests/helpers/test_recordhelper.py [hunk 10]
```diff
@@ -0,0 +1,104 @@
...
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program; if not, write to the Free Software
+# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+#
+#
+
```
Comment (frinzell) [inline tests/helpers/test_recordhelper.py:17]: Copyright year

File: tests/helpers/test_recordhelper.py [hunk 11]
```diff
@@ -0,0 +1,104 @@
...
+from filecmp import cmp
+from os import remove
+import unittest
+from os.path import join, isfile
+import chipsec.helper.replay.replayhelper as rph
+import chipsec.helper.record.recordhelper as rch
+
+
+class RecordHelperTest(unittest.TestCase):
+    # def setUp(self):
```
Comment (frinzell) [inline tests/helpers/test_recordhelper.py:30]: Dead code

File: tests/helpers/test_recordhelper.py [hunk 12]
```diff
@@ -0,0 +1,104 @@
...
+
+    def tearDown(self):
+        unittest.TestCase.tearDown(self)
+        if isfile(self.temp_file):
+            remove(self.temp_file)
+
+    def test_record(self):
+        # unittest.TestCase.setUp(self)
```
Comment (frinzell) [inline tests/helpers/test_recordhelper.py:38]: Dead code.

File: tests/helpers/test_replayhelper.py [hunk 13]
```diff
@@ -0,0 +1,105 @@
...
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program; if not, write to the Free Software
+# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+#
+#
```
Comment (frinzell) [inline tests/helpers/test_replayhelper.py:16]: Copyright year.

File: chipsec/helper/record/recordhelper.py (hunk 2 above)
Comment (npmitche) [reply chipsec/helper/record/recordhelper.py]: In my defense, there is no type hint for this function in the basehelper...

**Comment by npmitche** [comment]
Changes since (Aaron's) review: https://github.com/chipsec/chipsec/compare/8266f3f1d1df09acc312735b6762ac1265e80487..cb633341cb1414d1ce57a251959fcffa721eb2a8

File: chipsec/helper/record/recordhelper.py (hunk 2 above)
Comment (frinzell) [reply chipsec/helper/record/recordhelper.py]: Oops, guess we need to add that.  :)

**Review by dscott90 — APPROVED** [review]

**Review by frinzell — APPROVED** [review]

**Pull request merged**